README
-------------------------------------------------------------------------------
---  HIGHLIGHT MANUAL -------  Version 3.45   -------------- September 2018 ---
-------------------------------------------------------------------------------

OSI Certified Open Source Software

Deutsche Anleitung: README_DE

CONTENT
-------------------------------------------------------------------------------

1. OVERVIEW
   1.1 INTENDED PURPOSE
   1.2 FEATURE LIST
   1.3 SUPPORTED PROGRAMMING AND MARKUP LANGUAGES

2. USAGE AND OPTIONS
   2.1 QUICK INTRODUCTION
   2.2 CLI OPTIONS
   2.3 GUI OPTIONS
   2.4 INPUT AND OUTPUT
   2.5 GNU SOURCE-HIGHLIGHT COMPATIBILITY
   2.6 ADVANCED OPTIONS
   2.7 ENVIRONMENT VARIABLES
   2.8 SYNTAX TESTING

3. CONFIGURATION
   3.1 FILE FORMAT
   3.2 LANGUAGE DEFINITIONS 
   3.3 REGULAR EXPRESSIONS
   3.4 THEME DEFINITIONS
   3.5 KEYWORD GROUPS
   3.6 PLUG-INS
   3.7 FILE MAPPING
   3.8 CONFIG FILE SEARCH

4. EMBEDDING HIGHLIGHT
   4.1 SAMPLE SCRIPTS
   4.2 PANDOC
   4.3 SWIG
   4.4 TCL
   4.5 THIRD PARTY SCRIPTS AND PLUG-INS

5. BUILDING AND INSTALLING
   5.1 PRECOMPILED PACKAGES
   5.2 BUILDING DEPENDENCIES

6. DEVELOPER CONTACT


1. OVERVIEW
-------------------------------------------------------------------------------

Highlight converts sourcecode to HTML, XHTML, RTF, ODT, LaTeX, TeX, SVG, BBCode, 
Pango markup and terminal escape sequences with coloured syntax highlighting.
Syntax definitions and colour themes are customizable.


1.1 INTENDED PURPOSE
-------------------------------------------------------------------------------

Highlight was designed to offer a flexible but easy to use syntax highlighter
for several output formats. No syntax or colouring information is hardcoded,
instead all relevant data is stored in configuration scripts. These Lua scripts
may be altered and enhanced with plug-ins.


1.2 FEATURE LIST
-------------------------------------------------------------------------------

* highlighting of keywords, types, strings, numbers, escape sequences, comments,
  operators and preprocessor directives
* coloured output in HTML, XHTML 1.1, RTF, TeX, LaTeX, SVG, BBCode, Pango Markup 
  and terminal escape sequences
* supports referenced stylesheet files for HTML, LaTeX, TeX or SVG output
* configuration files are Lua scripts
* supports plug-in scripts to tweak language definitions and themes
* syntax elements are defined as regular expressions or plain string lists
* customizable keyword groups
* recognition of nested languages within a file
* reformatting and indentation of C, C++, C# and Java source code
* wrapping of long lines
* configurable output of line numbers


1.3 SUPPORTED PROGRAMMING AND MARKUP LANGUAGES
-------------------------------------------------------------------------------

Please see README_LANGLIST for the current set of supported languages.
You may also run "highlight --list-scripts=langs" to get a list and associated
file extensions.


2. USAGE AND OPTIONS
-------------------------------------------------------------------------------

2.1 QUICK INTRODUCTION
-------------------------------------------------------------------------------

The following examples show how to produce a highlighted C++ file, using 
main.cpp as input file:

- Generate HTML:
  highlight -i main.cpp -o main.cpp.html
  highlight < main.cpp > main.cpp.html --syntax cpp

  You will find the HTML file and highlight.css in the working directory.
  If you use IO redirection (2nd example), you must define the programming
  language with --syntax.

- Generate HTML with embedded CSS definitions and line numbers:
  highlight -i main.cpp -o main.cpp.html --include-style --line-numbers

- Generate HTML with inline CSS definitions:
  highlight -i main.cpp -o main.cpp.html --inline-css

- Generate LaTeX using "horstmann" source formatting style and "neon" colour
  theme: 
  highlight -O latex -i main.cpp -o main.cpp.tex --reformat horstmann --style neon

  The following output formats may be defined with --out-format:
  
  html:      HTML5 (default)
  xhtml:     XHTML 1.1
  tex:       Plain TeX
  latex:     LaTeX
  rtf:       RTF
  odt:       OpenDocument Text (Flat XML)
  svg:       SVG
  bbcode:    BBCode
  pango:     Pango markup
  ansi:      Terminal 16 color escape codes
  xterm256:  Terminal 256 color escape codes
  truecolor: Terminal 16m color escape codes

- Customize font settings:
  highlight --syntax ada --font-size 12 --font "'Courier New',monospace"
  highlight --syntax ada --out-format=latex --font-size tiny --font sffamily

- Define an output directory:
  highlight -d some/target/dir/ *.cpp *.h

See "highlight --help" or "man highlight" for more details.


2.2 CLI OPTIONS
-------------------------------------------------------------------------------

The command line version of highlight offers the following options:

USAGE: highlight [OPTIONS]... [FILES]...

General options:

 -B, --batch-recursive=<wc>     convert all matching files, searches subdirs
                                  (Example: -B '*.cpp')
 -D, --data-dir=<directory>     set path to data directory
     --config-file=<file>       set path to a lang or theme file
 -d, --outdir=<directory>       name of output directory
 -h, --help                     print this help
 -i, --input=<file>             name of single input file
 -o, --output=<file>            name of single output file
 -P, --progress                 print progress bar in batch mode
 -q, --quiet                    supress progress info in batch mode
 -S, --syntax=<type>            specify type of source code
 -v, --verbose                  print debug info
     --force                    generate output if input syntax is unknown
     --list-scripts=<type>      list installed scripts
                                  <type> = [langs, themes, plugins]
     --plug-in=<script>         execute Lua plug-in script; repeat option to
                                  execute multiple plug-ins
     --plug-in-param=<value>    set plug-in input parameter
     --print-config             print path configuration
     --print-style              print stylesheet only (see --style-outfile)
     --skip=<list>              ignore listed unknown file types
                                  (Example: --skip='bak;c~;h~')
     --start-nested=<lang>      define nested language which starts input
                                  without opening delimiter
     --stdout                   output to stdout (batch mode, --print-style)
     --validate-input           test if input is text, remove Unicode BOM
     --version                  print version and copyright information


Output formatting options:

 -O, --out-format=<format>      output file in given format
                                  <format>=[html, xhtml, latex, tex, odt, rtf,
                                  ansi, xterm256, truecolor, bbcode, pango, svg]
 -c, --style-outfile=<file>     name of style file or print to stdout, if
                                  'stdout' is given as file argument
 -e, --style-infile=<file>      to be included in style-outfile (deprecated)
                                  use a plug-in file instead
 -f, --fragment                 omit document header and footer
 -F, --reformat=<style>         reformats and indents output in given style
                                  <style> = [allman, banner, gnu,
                                  horstmann, java, kr, linux, mozilla, otbs, vtk,
                                  stroustrup, whitesmith, google, pico, lisp]
 -I, --include-style            include style definition in output file
 -J, --line-length=<num>        line length before wrapping (see -V, -W)
 -j, --line-number-length=<num> line number width incl. left padding (default: 5)
 -k, --font=<font>              set font (specific to output format)
 -K, --font-size=<num?>         set font size (specific to output format)
 -l, --line-numbers             print line numbers in output file
 -m, --line-number-start=<cnt>  start line numbering with cnt (assumes -l)
 -s, --style=<style>            set colour style (theme). See --base16
 -t, --replace-tabs=<num>       replace tabs by <num> spaces
 -T, --doc-title=<title>        document title
 -u, --encoding=<enc>           set output encoding which matches input file
                                  encoding; omit encoding info if set to NONE
 -V, --wrap-simple              wrap lines after 80 (default) characters w/o
                                  indenting function parameters and statements
 -W, --wrap                     wrap lines after 80 (default) characters
     --wrap-no-numbers          omit line numbers of wrapped lines
                                  (assumes -l)
 -z, --zeroes                   pad line numbers with 0's
     --base16                   use a theme of the Base16 collection
     --delim-cr                 set CR as end-of-line delimiter (MacOS 9)
     --keep-injections          output plug-in injections in spite of -f
     --kw-case=<case>           change case of case insensitive keywords
                                  <case> =  [upper, lower, capitalize]
     --no-trailing-nl           omit trailing newline
     --no-version-info          omit version info comment


(X)HTML output options:

 -a, --anchors                  attach anchor to line numbers
 -y, --anchor-prefix=<str>      set anchor name prefix
 -N, --anchor-filename          use input file name as anchor prefix
 -C, --print-index              print index with hyperlinks to output files
 -n, --ordered-list             print lines as ordered list items
     --class-name=<name>        set CSS class name prefix;
                                  omit class name if set to NONE
     --inline-css               output CSS within each tag (verbose output)
     --enclose-pre              enclose fragmented output with pre tag 
                                  (assumes -f)


LaTeX output options:

 -b, --babel                    disable Babel package shorthands
 -r, --replace-quotes           replace double quotes by \dq{}
     --beamer                   adapt output for the Beamer package
     --pretty-symbols           improve appearance of brackets and other symbols


RTF output options:

     --page-color               include page color attributes
 -x, --page-size=<ps>           set page size 
                                  <ps> = [a3, a4, a5, b4, b5, b6, letter]
     --char-styles              include character stylesheets


SVG output options:

     --height                   set image height (units allowed)
     --width                    set image width (see --height)


GNU source-highlight compatibility options:

     --doc                      create stand alone document
     --no-doc                   cancel the --doc option
     --css=filename             the external style sheet filename
     --src-lang=STRING          source language
 -t, --tab=INT                  specify tab length
 -n, --line-number[=0]          number all output lines, optional padding
     --line-number-ref[=p]      number all output lines and generate an anchor,
                                  made of the specified prefix p + the line
                                  number  (default='line')
     --output-dir=path          output directory
     --failsafe                 if no language definition is found for the
                                  input, it is simply copied to the output
                                  
2.3 GUI OPTIONS
-------------------------------------------------------------------------------

The Graphical User Interface offers a subset of the CLI's features. It includes
a dynamic preview of the output file's apperarance. Please see screenshots and
screencasts on the project website.
Invoke highlight-gui with the --portable option to let it save its settings
in the binary's current directory (instead of using the registry).


2.4 INPUT AND OUTPUT
-------------------------------------------------------------------------------

If no input or output file name is defined by --input and --output options,
highlight will use stdin and stdout for file processing.
Since version 3.44, reading from stdin can also be triggered by the "-" option.

If no input filename is defined by --input or given at the prompt, highlight is
not able to determine the language type by means of the file extension (except
some scripting languages which are figured out by the shebang in the first input
line). In this case you have to pass highlight the language with --syntax
(this usually should be the file suffix of the source file).
Example: If you want to convert a Python file, highlight needs to load the
py.lang definition. The correct argument of --syntax would be "py".

highlight test.py
highlight < test.py --syntax py       # --syntax option necessary
cat test.py | highlight --syntax py

If there exist multiple suffixes (like C, cc, cpp, h for C++ - files),
they are mapped to a language definition in $CONF_DIR/filetypes.conf.

Highlight enters the batch processing mode if multiple input files are given
or if --batch-recursive is set.
In batch mode, highlight will save the generated files using the original
filename, appending the extension of the chosen output type.
If files in the input directories happen to share the same name, the output
files will be prefixed with their source path name.
The --out-dir option is recommended in batch mode. Use --quiet to improve
performance (recommended for usage in shell scripts).

HTML, TeX, LaTeX and SVG output
-------------------------------

The HTML, TeX, LaTeX and SVG output formats allow to reference a stylesheet
file which contains the formatting information.

In HTML and SVG output, this file contains CSS definitions and is saved as
'highlight.css'. In LaTeX and TeX, it contains macro definitions, and is saved
as 'highlight.sty'.

Name and path of the stylesheet may be modified with --style-outfile.
If the --outdir option is given, all generated output, including stylesheets,
are stored in this directory.

Use --include-style to embed the style information in the output documents
without referencing a stylesheet.

Referenced stylesheets have the advantage to share all formatting information
in a single file, which affects all referencing documents.

With --style-infile you define a file to be included in the final formatting
information of the document. This way you enhance or redefine the default
highlight style definitions without editing generated code.
Note: Using a plug-in script is the preferred way to enhance styling.

Terminal output:
----------------

Since there are limited colours defined for ANSI terminal output, there exists
only one hard coded colour theme with --out-format=ansi. You should therefore
use --out-format=xterm256 to enable output in 256 colours. The 256 colour mode
is supported by recent releases of xterm, rxvt and Putty (among others).
The latest terminal emulators also support 16m colors, this mode is enabled
with --out-format=truecolors.

highlight --out-format=ansi <inputfile> | less -R
highlight --out-format=xterm256 <inputfile> | less -R

Text processing:
----------------

If the language definition is specified as "txt", no highlighting takes place.

highlight -S txt --out-format=latex README > README.tex


2.5 GNU SOURCE-HIGHLIGHT COMPATIBILITY
-------------------------------------------------------------------------------

The command line interface is extensively harmonised with source-highlight
(http://www.gnu.org/software/src-highlite/).

The following highlight options have the same meaning as in source-highlight:
 --input, --output, --help, --version, --out-format, --title, --data-dir,
 --verbose, --quiet

These options were added to enhance compatibility:
 --css, --doc, --failsafe, --line-number, --line-number-ref, --no-doc, --tab,
 --output-dir, --src-lang

These switches provide a common highlighter interface for scripts, plugins etc.


2.6 ADVANCED OPTIONS
-------------------------------------------------------------------------------

Prevent parsing of binary input files
-------------------------------------

If highlight might process untrusted input, you can disable parsing of binary 
files using --validate-input. This flag causes highlight to match the input file 
header with a list of magic numbers. If a binary file type is detected, highlight 
quits with an error message. This switch also removes an UTF-8 BOM in the output.

Highlight nested code without starting delimiter
------------------------------------------------

If a file starts with an embedded code section which misses an appropriate opening
delimiter, the --start-nested option will switch to the nested language mode.
This can be useful with LuaTeX files:
highlight luatex.tex --latex --start-nested=inc_luatex

inc_luatex is a Lua language definition with TeX line comments.
The nested code section has to end with the ending delimiter defined in the host 
language definition.

Test new configuration scripts
------------------------------

The option --config-file helps to test new config files. The argument file must be
a lang or theme.

highlight --config-file xxx.lang --config-file yyy.theme -I

Debug language definitions
--------------------------

Use --verbose to display Lua and syntax data.

Remove an UTF8 BOM:
-------------------

Use --validate-input to get rid of UTF8 byte order marks.

Force output to stdout
----------------------

Use --stdout to write output files in batch mode to stdout.

Portable GUI (Windows build)
----------------------------

Invoke highlight-gui.exe with the --portable switch to save its configuration
in text files instead of the registry.


2.7 ENVIRONMENT VARIABLES
-------------------------------------------------------------------------------

The command line version recognizes these variables:

HIGHLIGHT_DATADIR: sets the path to highlight's configuration scripts
HIGHLIGHT_OPTIONS: may contain command line options, but no input file paths.

2.8 SYNTAX TESTING
-------------------------------------------------------------------------------

Since version 2.45, highlight supports special notations within comments to
test its syntax recognition. See README_TESTCASES for details.


3. CONFIGURATION
-------------------------------------------------------------------------------

3.1 FILE FORMAT
-------------------------------------------------------------------------------

Configuration files are Lua scripts. 
Please refer to http://www.lua.org/manual/5.1/manual.html for more details
about the Lua syntax.

These constructs are sufficient to edit the scripts:

Variable assigment:
name = value
(variables have no type, only values have)

Strings:

ChangeLog
CHANGELOG Highlight


highlight 3.47

xx.10.2018

-fixed xterm256 and truecolor whitespace output #2 (https://gitlab.com/saalen/highlight/issues/90)

---
highlight 3.46

07.10.2018

-fixed xterm256 and truecolor whitespace output (https://gitlab.com/saalen/highlight/issues/90)
-converted manuals to AsciiDoc (thanks to Tristano Ajmone)

---
highlight 3.45

02.10.2018

-added DocumentHeader and DocumentFooter plug-in hooks
-added RemoveKeyword Lua function for syntax definitions
-added syntax test indicators (see README_TESTCASES)
-added support for ISO and R10 variants of Modula2 (thanks to Benjamin Kowarsch)
-fixed R identifiers (https://gitlab.com/saalen/highlight/merge_requests/77)
-fixed ALAN IF identifiers (see https://gitlab.com/saalen/highlight/merge_requests/79)
-fixed issue with Bash string interpolation
-fixed some bugs found by coverity tool (https://gitlab.com/saalen/highlight/issues/82)
-added Swift keywords and types
-added filetypes.conf.5 man page (https://gitlab.com/saalen/highlight/issues/83)
-added Gradle extension mapping (https://gitlab.com/saalen/highlight/merge_requests/80)

---
highlight 3.44

17.07.2018

-fixed gcc 8 compilation warnings
-fixed Ruby string interpolation 
  (https://github.com/andre-simon/highlight/issues/70)
-added support for ALAN IF (thanks to Tristano Ajmone)
-added 107 Base16 themes [https://github.com/chriskempson/base16] 
  (thanks to Tristano Ajmone)
-updated Rust and Java reserved words lists
-revised documentation
-moved extras/css-themes into extras/themes-resources
-added extras/themes-resources/base16 (thanks to Tristano Ajmone)
-GUI: added Base16 theme selection checkbox
-CLI: added --base16 option to enable the new themes
-CLI: accept - as argument to read from stdin 
  (https://github.com/andre-simon/highlight/issues/73)

---
highlight 3.43

30.04.2018

-updated astyle code to release 3.1 (Rev. 655)
-added webkit reformatting style
-improved several language definitions
-fixed Matlab string recognition 
  (https://github.com/andre-simon/highlight/issues/61)
-fixed Autohotkey escape sequence recognition (thanks to Klaus Daube)
-added excel.lang (https://github.com/andre-simon/highlight/pull/60)
-improved Qt pro file (https://github.com/andre-simon/highlight/pull/59)
-CLI: added --reformat-option 
  (https://github.com/andre-simon/highlight/issues/62)
-CLI: added --line-range (https://github.com/andre-simon/highlight/issues/64)
-GUI: added Bulgarian translation  (thanks to Georgi Sotirov)

---
highlight 3.42

20.01.2018

-fixed HL_OUTPUT in Lua state for HL_FORMAT_XHTML and HL_FORMAT_TRUECOLOR values
-fixed lib-shared make target
-updated astyle code to release 3.1.0 beta
-added Polygen and EBNF2 syntax definitions (thanks to Tristano Ajmone)
-added pywal terminal colouring template in extras/pywal
-added reformatting style ratliff (replaces banner)
-added extras/langDefs-resources/cleanslate.lang (thanks to Tristano Ajmone)
-improved Perl6 compatibility
-improved PHP string interpolation
-improved Haskell definition (https://github.com/andre-simon/highlight/pull/52)
-CLI: added --canvas option to define background color padding in ANSI output
  (https://github.com/andre-simon/highlight/issues/40)
-GUI: added French translation (thanks to Antoine Belvire) 
-GUI: added Scripts tab (suggested by Tristano Ajmone) 
-GUI: minor bugfixes

---
highlight 3.41

27.11.2017

-renamed examples directory to extras
-line anchors (-a) are attached as id attribute to the first span or li tag in 
  HTML output (https://github.com/andre-simon/highlight/issues/36)
-renamed ID prefix in outhtml_codefold plug-in to be compatible with -a IDs
-added fstab.lang and added anacrontab in filetypes.conf
-removed references to OutputType::HTML32
-added extras/css-themes and extras/langDefs-resources
  (thanks to Tristano Ajmone)
-CLI: removed deprecated indicator of --data-dir option
-CLI: added --no-version-info option
-GUI: fixed initial theme selection
-GUI: added "Omit version info comment" option
-GUI: added "Copy with MIME type" option for HTML output 
  (https://github.com/andre-simon/highlight/issues/32) 

---
highlight 3.40

20.10.2017

-fixed Ruby string parsing (thanks to Jens Schleusener)
-fixed segfault on sparc64 (patch by James Clarke)
-fixed PureBasic definition (https://github.com/andre-simon/highlight/issues/25)
-added CMake definition (https://github.com/andre-simon/highlight/issues/20)
-added email definition (https://github.com/andre-simon/highlight/issues/21)
-linked scm suffix to lisp definition 
  (https://github.com/andre-simon/highlight/issues/22)
-W32 CLI: support HIGHLIGHT_DATADIR and --data-dir options 
  (https://github.com/andre-simon/highlight/issues/24)
-revised documentation

---
highlight 3.39

25.07.2017

-added syntax for Docker and Elixir
-improved HTML, Julia, Kotlin and Smalltalk syntax definitions
-GUI: added "Paste, Convert and Copy" button 
  (https://sourceforge.net/p/syntaxhighlight/support-requests/4/)

---
highlight 3.38

20.06.2017

-fixed Bash variable highlighting issue
-updated astyle code to release 3.0.1 (https://sourceforge.net/p/astyle/bugs/438)
-added bash_ref_man7_org.lua plugin

---
highlight 3.37

30.05.2017

-fixed Perl string highlighting issue
-fixed highlighting if a line continues after the nested code delimiter
-updated astyle code to release 3.0
-added examples/pandoc (thanks to Tristano Ajmone)
-added syntax mapping for markdown 
  (https://github.com/andre-simon/highlight/issues/11)
-added syntax mapping for clj 
  (https://github.com/andre-simon/highlight/issues/15)
-improved Java definition (https://github.com/andre-simon/highlight/issues/13)
-added theme to JSON converter in  examples/json 
  (https://github.com/andre-simon/highlight/issues/8)
-CLI: added support for environment variable HIGHLIGHT_OPTIONS 
  (https://github.com/andre-simon/highlight/issues/17)

---
highlight 3.36

30.03.2017

-fixed code folding plugin to support more Ruby conditional modifiers 
  (thanks to Jens Schleusener)

src/cli/help.cpp
/***************************************************************************
                          help.cpp  -  description
                             -------------------
    begin                : Die Apr 23 2002
    copyright            : (C) 2002-2018 by Andre Simon
    email                : andre.simon1@gmx.de
 ***************************************************************************/

/*
This file is part of Highlight.

Highlight is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Highlight is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Highlight.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include "help.h"

using namespace std;

namespace Help
{
void printHelp()
{
    cout<<"USAGE: highlight [OPTIONS]... [FILES]...\n";
    cout<<"\n";
    cout<<"General options:\n";
    cout<<"\n";
    cout<<" -B, --batch-recursive=<wc>     convert all matching files, searches subdirs\n";
    cout<<"                                  (Example: -B '*.cpp')\n";
    cout<<" -D, --data-dir=<directory>     set path to data directory\n";
    cout<<"     --config-file=<file>       set path to a lang or theme file\n";
    cout<<" -d, --outdir=<directory>       name of output directory\n";
    cout<<" -h, --help                     print this help\n";
    cout<<" -i, --input=<file>             name of single input file\n";
    cout<<" -o, --output=<file>            name of single output file\n";
    cout<<" -P, --progress                 print progress bar in batch mode\n";
    cout<<" -q, --quiet                    supress progress info in batch mode\n";
    cout<<" -S, --syntax=<type>            specify type of source code\n";
    cout<<" -v, --verbose                  print debug info\n";
    cout<<"     --force                    generate output if input syntax is unknown\n";
    cout<<"     --list-scripts=<type>      list installed scripts\n";
    cout<<"                                  <type> = [langs, themes, plugins]\n";
    cout<<"     --plug-in=<script>         execute Lua plug-in script; repeat option to\n";
    cout<<"                                  execute multiple plug-ins\n";
    cout<<"     --plug-in-param=<value>    set plug-in input parameter\n";
    cout<<"     --print-config             print path configuration\n";
    cout<<"     --print-style              print stylesheet only (see --style-outfile)\n";
    cout<<"     --skip=<list>              ignore listed unknown file types\n";
    cout<<"                                  (Example: --skip='bak;c~;h~')\n";
    cout<<"     --start-nested=<lang>      define nested language which starts input\n";
    cout<<"                                  without opening delimiter\n";
    cout<<"     --stdout                   output to stdout (batch mode, --print-style)\n";
    cout<<"     --validate-input           test if input is text, remove Unicode BOM\n";
    cout<<"     --version                  print version and copyright information\n";
    cout<<"\n";
    cout<<"\n";
    cout<<"Output formatting options:\n";
    cout<<"\n";
    cout<<" -O, --out-format=<format>      output file in given format\n";
    cout<<"                                  <format>=[html, xhtml, latex, tex, odt, rtf,\n";
    cout<<"                                  ansi, xterm256, truecolor, bbcode, pango, svg]\n";
    cout<<" -c, --style-outfile=<file>     name of style file or print to stdout, if\n";
    cout<<"                                  'stdout' is given as file argument\n";

    cout<<" -e, --style-infile=<file>      to be included in style-outfile (deprecated)\n";
    cout<<"                                  use a plug-in file instead\n";

    cout<<" -f, --fragment                 omit document header and footer\n";
    cout<<" -F, --reformat=<style>         reformats and indents output in given style\n";
    cout<<"                                  <style> = [allman, gnu, google, horstmann,\n";
    cout<<"                                  java, kr, linux, lisp, mozilla, otbs, pico,\n";
    cout<<"                                  vtk, ratliff, stroustrup, webkit, whitesmith, user]\n";
    cout<<"                                  The user style does not apply a predefined scheme.\n";
    cout<<"                                  Use --reformat-option to define the reformatting.\n";
    cout<<"     --reformat-option=<opt>    apply an astyle cmd line option (assumes -F)\n";
    cout<<" -I, --include-style            include style definition in output file\n";
    cout<<" -J, --line-length=<num>        line length before wrapping (see -V, -W)\n";
    cout<<" -j, --line-number-length=<num> line number width incl. left padding (default: 5)\n";
    cout<<"     --line-range=<start-end>   output only lines from number <start> to <end>\n";

    cout<<" -k, --font=<font>              set font (specific to output format)\n";
    cout<<" -K, --font-size=<num?>         set font size (specific to output format)\n";
    cout<<" -l, --line-numbers             print line numbers in output file\n";
    cout<<" -m, --line-number-start=<cnt>  start line numbering with cnt (assumes -l)\n";
    cout<<" -s, --style=<style>            set colour style (theme). See  --base16\n";
    cout<<" -t, --replace-tabs=<num>       replace tabs by <num> spaces\n";
    cout<<" -T, --doc-title=<title>        document title\n";
    cout<<" -u, --encoding=<enc>           set output encoding which matches input file\n";
    cout<<"                                  encoding; omit encoding info if set to NONE\n";
    cout<<" -V, --wrap-simple              wrap lines after 80 (default) characters w/o\n";
    cout<<"                                  indenting function parameters and statements\n";
    cout<<" -W, --wrap                     wrap lines after 80 (default) characters\n";
    cout<<"     --wrap-no-numbers          omit line numbers of wrapped lines\n";
    cout<<"                                  (assumes -l)\n";
    cout<<" -z, --zeroes                   pad line numbers with 0's\n";
    cout<<"     --base16                   use a theme of the Base16 collection\n";
    cout<<"     --delim-cr                 set CR as end-of-line delimiter (MacOS 9)\n";
    cout<<"     --keep-injections          output plug-in injections in spite of -f\n";
    cout<<"     --kw-case=<case>           change case of case insensitive keywords\n";
    cout<<"                                  <case> =  [upper, lower, capitalize]\n";
    cout<<"     --no-trailing-nl           omit trailing newline\n";
    cout<<"     --no-version-info          omit version info comment\n";
    cout<<"\n\n";
    cout<<"(X)HTML output options:\n";
    cout<<"\n";
    cout<<" -a, --anchors                  attach anchor to line numbers\n";
    cout<<" -y, --anchor-prefix=<str>      set anchor name prefix\n";
    cout<<" -N, --anchor-filename          use input file name as anchor prefix\n";
    cout<<" -C, --print-index              print index with hyperlinks to output files\n";
    cout<<" -n, --ordered-list             print lines as ordered list items\n";
    cout<<"     --class-name=<name>        set CSS class name prefix;\n";
    cout<<"                                  omit class name if set to NONE\n";
    cout<<"     --inline-css               output CSS within each tag (verbose output)\n";
    cout<<"     --enclose-pre              enclose fragmented output with pre tag \n";
    cout<<"                                  (assumes -f)\n";
    cout<<"\n\n";
    cout<<"LaTeX output options:\n";
    cout<<"\n";
    cout<<" -b, --babel                    disable Babel package shorthands\n";
    cout<<" -r, --replace-quotes           replace double quotes by \\dq{}\n";
    cout<<"     --beamer                   adapt output for the Beamer package\n";
    cout<<"     --pretty-symbols           improve appearance of brackets and other symbols\n";
    cout<<"\n";
    cout<<"\n";
    cout<<"RTF output options:\n";
    cout<<"\n";
    cout<<"     --page-color               include page color attributes\n";
    cout<<" -x, --page-size=<ps>           set page size \n";
    cout<<"                                  <ps> = [a3, a4, a5, b4, b5, b6, letter]\n";
    cout<<"     --char-styles              include character stylesheets\n";
    cout<<"\n";
    cout<<"\n";
    cout<<"SVG output options:\n";
    cout<<"\n";
    cout<<"     --height                   set image height (units allowed)\n";
    cout<<"     --width                    set image width (see --height)\n";
    cout<<"\n";
    cout<<"\n";
    cout<<"Terminal escape output options (xterm256 or truecolor):\n";
    cout<<"\n";
    cout<<"     --canvas[=width]           set background colour padding (default: 80)\n";
    cout<<"\n";
    cout<<"\n";
    cout<<"GNU source-highlight compatibility options:\n";
    cout<<"\n";
    cout<<"     --doc                      create stand alone document\n";
    cout<<"     --no-doc                   cancel the --doc option\n";
    cout<<"     --css=filename             the external style sheet filename\n";
    cout<<"     --src-lang=STRING          source language\n";
    cout<<" -t, --tab=INT                  specify tab length\n";
    cout<<" -n, --line-number[=0]          number all output lines, optional padding\n";
    cout<<"     --line-number-ref[=p]      number all output lines and generate an anchor,\n";
    cout<<"                                  made of the specified prefix p + the line\n";
    cout<<"                                  number  (default='line')\n";
    cout<<"     --output-dir=path          output directory\n";
    cout<<"     --failsafe                 if no language definition is found for the\n";
    cout<<"                                  input, it is simply copied to the output\n";
    cout<<"\n";
    cout<<"\n";
    cout<<"If no in- or output files are specified, stdin and stdout will be used.\n";
    cout<<"Since version 3.44, reading from stdin can also be triggered by the '-' option.\n";
    cout<<"Default output format: xterm256 or truecolor if appropriate, HTML otherwise.\n";
    cout<<"Style definitions are stored in highlight.css (HTML, XHTML, SVG) or\n";
    cout<<"highlight.sty (LaTeX, TeX) if neither -c nor -I is given.\n";
    cout<<"Reformatting code (-F) will only work with C, C++, C# and Java input files.\n";
    cout<<"Wrapping lines with -V or -W will cause faulty highlighting of long single\n";
    cout<<"line comments and directives. Using line-range might interfere with multi\n";
    cout<<"line syntax elements. Use with caution.\n\n";
    cout<<"Run highlight --list-scripts=langs to see all supported syntax types.\n\n";
    cout<<"Refer to README files how to apply plug-ins to customize the output.\n\n";
    cout<<"Updates and information: http://www.andre-simon.de/\n";
}

}
